lib/sda/service.go

```diff
@@ -0,0 +1,387 @@
+// Package sda has the service.go file which contains the code that glues the innerpart of SDA with the external
+// part. Basically, you have the definition of a Protocol which anybody can
+// implement, this is what sda is for: to run protocols.
+//
+// A Protocol in sda is supposed to be a short term struct that runs its
+// algorithm and then finishes. SDA handles the rounds and the dispatching.
+// Multiple protocols can run in parallel with the same Tree / EntityList etc.
+// You  can registers Protocol at init time using `sda.ProtocolRegister`.
+// If you need a long term object to store longterm information that Protocol
+// can use (for example, the hash of the last signature a given protocol has
+// issued to create a blockchain), you should implement the Service interface.
+//
+// A service is a longterm running platform that performs two main functions:
+// * First, it creates the ProtocolInstances out of a Node. SDA will request the
+// right service each time it needs to create a new one. The service has to
+// provide any external additional information to the protocol so it can work
+// knowing everything it needs to know.
+// * Secondly, a Service responds to requests made by clients through a CLI
+// using the API. The service will be responsible to create a ProtocolInstance
+// out of the client's request with the right Tree + EntityList and starts the
+// protocol.
+// The service can reply to requests coming also from others services or others
+// nodes; the Request is only a wrapper around JSON object.
+// All services are registered at init time and directly startup in the main
+// function.
+package sda
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/dedis/cothority/lib/dbg"
+	"github.com/dedis/cothority/lib/network"
+	"github.com/satori/go.uuid"
+	"os"
+	"path"
+)
+
+// ProtocolInstance is the interface that instances have to use in order to be
+// recognized as protocols
+type ProtocolInstance interface {
+	// Start is called when a leader has created its tree configuration and
+	// wants to start a protocol, it calls host.StartProtocol(protocolID), that
+	// in turns instantiate a new protocol (with a fresh token), and then call
+	// Start on it.
+	Start() error
+	// Dispatch is called as a go-routine and can be used to handle channels
+	Dispatch() error
+	// Shutdown cleans up the resources used by this protocol instance
+	Shutdown() error
+}
+
+// NewProtocol is a convenience to represent the cosntructor function of a
+// ProtocolInstance
+type NewProtocol func(*Node) (ProtocolInstance, error)
+
+// ProtocolConstructor is an interface that can instantiate a given protocol
+type ProtocolConstructor interface {
+	NewProtocol(*Node) (ProtocolInstance, error)
+}
+
+// Service is a generic interface to define any type of services.
+type Service interface {
```

Comment:
Should loading and saving of configuration-variables be in here? Something like

```
Load(file string) error
Save(file string) error
```

Where the `sda` provides a filename to the service that can be used to save the configuration? It could also be a directory where the service can store more than one file.